File: apps/patter-voice-sdk/docs/examples/tool-calling.py
"""Agent with tool calling — checks inventory during the call."""
import asyncio
from getpatter import Patter, Twilio, OpenAIRealtime, Tool


async def main():
    phone = Patter(
        carrier=Twilio(),                               # TWILIO_* from env
        phone_number="+15550001234",
        webhook_url="xxx.ngrok-free.dev",
    )

    agent = phone.agent(
        engine=OpenAIRealtime(),                        # OPENAI_API_KEY from env
        system_prompt="You help customers check product availability.",
        tools=[
            Tool(
                name="check_stock",
                description="Check if a product is in stock",
                parameters={"type": "object", "properties": {"product": {"type": "string"}}},
                webhook_url="https://api.example.com/stock",
            ),
        ],
    )

    print("Listening for calls...")
    await phone.serve(agent=agent, port=8000)


asyncio.run(main())

File: apps/patter-voice-sdk/docs/examples/tool-calling.ts
/**
 * Tool calling — agent checks inventory during the call.
 *
 * Tools are defined with a name, description, JSON schema, and a webhookUrl.
 * Patter calls the webhook when the agent decides to use the tool and feeds
 * the result back into the conversation.
 */
import { createServer } from "http";
import { Patter, Twilio, OpenAIRealtime } from "getpatter";

// Mock inventory webhook — replace with your real endpoint
const toolServer = createServer((req, res) => {
  let body = "";
  req.on("data", (chunk) => (body += chunk));
  req.on("end", () => {
    const data: { arguments?: { product?: string } } = JSON.parse(body);
    const stock: Record<string, boolean> = { widget: true, gadget: false };
    const product = data.arguments?.product ?? "";
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ in_stock: stock[product] ?? false }));
  });
});
toolServer.listen(8001, () => console.log("Tool webhook listening on :8001"));

const phone = new Patter({
  carrier: new Twilio(),                                // TWILIO_* from env
  phoneNumber: "+15550001234",
  webhookUrl: "xxx.ngrok-free.dev",
});

const agent = phone.agent({
  engine: new OpenAIRealtime({ voice: "alloy" }),       // OPENAI_API_KEY from env
  systemPrompt:
    "You help customers check product availability. " +
    "Always use the check_stock tool before answering stock questions.",
  firstMessage: "Hello! I can help you check if a product is in stock. What are you looking for?",
  tools: [
    {
      name: "check_stock",
      description: "Check if a product is in stock",
      parameters: {
        type: "object",
        properties: { product: { type: "string", description: "Product name to look up" } },
        required: ["product"],
      },
      webhookUrl: "http://localhost:8001/stock",
    },
  ],
});

console.log("Listening for calls...");
await phone.serve({ agent, port: 8000 });
